codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServiceConfigGenerator.kt

```diff
@@ -125,20 +243,201 @@ class ServiceConfigGenerator(
+            "BuilderBuildMethod" to builderBuildMethod(),
+        )
+    }
+
+    private fun isBuilderFallible() = configMethods.isBuilderFallible()
+
+    private fun builderBuildRequiredMethodChecks() = configMethods.filter { it.isRequired }.map {
+        writable {
+            rust(
+                """
+                if !self.${it.requiredBuilderFlagName()} {
+                    return Err(${serviceName}ConfigError::${it.requiredErrorVariant()});
+                }
+                """,
+            )
+        }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsDefinitions() = configMethods.filter { it.isRequired }.map {
+        writable { rust("pub(crate) ${it.requiredBuilderFlagName()}: bool,") }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsInit() = configMethods.filter { it.isRequired }.map {
+        writable { rust("${it.requiredBuilderFlagName()}: false,") }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsMove() = configMethods.filter { it.isRequired }.map {
+        writable { rust("${it.requiredBuilderFlagName()}: self.${it.requiredBuilderFlagName()},") }
+    }.join("\n")
+
+    private fun builderRequiredMethodError() = writable {
+        val variants = configMethods.filter { it.isRequired }.map {
+            writable {
+                rust(
+                    """
+                    ##[error("service is not fully configured; invoke `${it.requiredBuilderFlagName()}` on the config builder")]
+                    ${it.requiredErrorVariant()},
+                    """,
+                )
+            }
+        }
+        if (isBuilderFallible()) {
+            rustTemplate(
+                """
+                ##[derive(Debug, #{ThisError}::Error)]
+                pub enum ${serviceName}ConfigError {
+                    #{Variants:W}
+                }
+                """,
+                "ThisError" to ServerCargoDependency.ThisError.toType(),
+                "Variants" to variants.join("\n"),
+            )
+        }
+    }
+
+    private fun injectedMethods() = configMethods.map {
+        writable {
+            val paramBindings = it.params.map { binding ->
+                writable { rustTemplate("${binding.name}: #{BindingTy},", "BindingTy" to binding.ty) }
+            }.join("\n")
+
+            // This produces a nested type like: "S<B, S<B, T>>", where
+            // - "S" denotes a "stack type" with two generic type parameters: the first is the top of the stack and the
+            //   second is the rest of the stack. For example, `aws_smithy_http_server::plugin::PluginStack`.
+            // - "B" is the type of the "thing" that is added.
+            // - "T" is the generic type variable name used in the enclosing impl block.
+            fun List<Binding>.stackReturnType(genericTypeVarName: String, stackType: RuntimeType): Writable =
+                this.fold(writable { rust(genericTypeVarName) }) { acc, next ->
+                    writable {
```

Comment:
nit: So, if I return a `listOf(A, B, C)`, does it get transformed into `Stack<C, Stack<B, Stack<A, T>>>`, where `T` is `genericTypeVarName`? If yes, how about we use `this.foldRight` so that it becomes `Stack<A, Stack<B, Stack<C, T>>>`?

Reply:
The inner middleware is the left type parameter of the stack: https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html

So we do want `Stack<C, Stack<B, Stack<A, T>>>`. A will be executed first, then B, then C.

We should document this though.